GVFS/GVFS.Service/ProductUpgradeTimer.cs

```diff
@@ -53,70 +52,101 @@ public void Dispose()
-                    // upgraded by manually downloading and running asset installers.
-                    ProductUpgraderInfo.DeleteAllInstallerDownloads();
-                    errorMessage = null;
-                    return true;
-                }
+                        return;
+                    }
 
-                if (productUpgrader.TryDownloadNewestVersion(out detailedError))
-                {
-                    errorMessage = null;
-                    return true;
+                    InstallerPreRunChecker prerunChecker = new InstallerPreRunChecker(this.tracer, string.Empty);
+                    if (!prerunChecker.TryRunPreUpgradeChecks(out _))
+                    {
+                        return;
+                    }
+
+                    if (!productUpgrader.UpgradeAllowed(out errorMessage))
+                    {
+                        if (errorMessage != null)
+                        {
+                            this.tracer.RelatedError(errorMessage);
+                        }
+
+                        return;
+                    }
+
+                    if (!this.TryQueryForNewerVersion(
+                            activity,
+                            productUpgrader,
+                            out Version newerVersion,
+                            out errorMessage))
+                    {
+                        if (errorMessage != null)
+                        {
+                            this.tracer.RelatedError(errorMessage);
+                        }
+
+                        return;
+                    }
+
+                    ProductUpgraderInfo.RecordHighestAvailableVersion(newerVersion);
                 }
-                else
+                catch (Exception ex) when (
+                    ex is IOException ||
+                    ex is UnauthorizedAccessException ||
+                    ex is NotSupportedException)
                 {
-                    errorMessage = "Could not download product upgrade. " + detailedError;
-                    return false;
+                    this.tracer.RelatedError(
+                        CreateEventMetadata(ex),
+                        "Exception encountered recording highest available version");
                 }
             }
         }
+
+        private bool TryQueryForNewerVersion(ITracer tracer, GitHubUpgrader productUpgrader, out Version newVersion, out string errorMessage)
+        {
+            errorMessage = null;
+            tracer.RelatedInfo("Querying server for latest version...");
+
+            if (!productUpgrader.TryQueryNewestVersion(out newVersion, out string detailedError))
+            {
+                errorMessage = "Could not fetch new version info. " + detailedError;
+                return false;
+            }
+
+            string logMessage = newVersion is null ? "No newer versions available." : $"Newer version available: {newVersion}.";

```

Comment:
Minor: I think this came up in another PR as well, but elsewhere in the codebase we're using ` == null` to check if a variable is `null` and we should stay consistent.